pandas/_libs/lib.pyx

```diff
@@ -1356,6 +1380,29 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
+        """
+        Gets the current seen val and casts it to ret_val's type.
+        Caller is responsible for making sure that the cast is safe
+        """
+        if self.curr_seen is CurrSeen.bool_t:
+            ret_val[0] = <curr_seen_t>self.curr_seen_val.bool_t
```

Comment:
Instead of this is there a way to convert this into a class that can manage the state of the known type as well as the corresponding data? I think if you go with a class-based approach you can more naturally achieve the polymorphic behavior you are ultimately after, but right now its scattered across the implementation.

Reply:
In the latest commit, I've moved handling of setting item values to the Seen object, so right now everything is completely encapsulated within the Seen class. Is this what you were thinking of?